### ISSUE TITLE
Gogs is Serving Web Resources From the Docker Gateway IP Instead of the Docker Host IP

### ISSUE THREAD

**Issue by oleersoy** [gogits/gogs#3019]
Hi,

I'm running the gogs docker installation on Ubuntu 14.04, Docker version 1.11.0, build 4dc5990, and SQLite using these instructions:

https://github.com/gogits/gogs/tree/master/docker

After I run the command:
`ole@MKI:~$ docker run --name=gogs -p 10022:22 -p 10080:3000 -v gogs-data:/data gogs/gogs`

I run:

```
ole@MKI:~$ docker inspect gogs | grep IPAddress
            "IPAddress": "172.17.0.3",
                    "IPAddress": "172.17.0.3",
```

To get the IPAddress assigned to the gogs docker container.  I then use this IPAddress per the installation instructions.  This results in gogs redirecting to:

`http://172.17.0.3:10080/user/login` which has the standard `Unable to connect` message in firefox.  

Gogs does respond to the Docker gateway ip address though.  So if I change the login url to:
`http://172.17.0.1:10080/user/login` then I can connect, create a user account, and login.

Here is the entire install trace that gogs prints to the console while the install is going on:
ole@MKI:~$ docker run --name=gogs -p 10022:22 -p 10080:3000 -v gogs-data:/data gogs/gogs
Apr 25 15:26:22 syslogd started: BusyBox v1.24.1
2016/04/25 15:26:22 [W] Custom config '/data/gogs/conf/app.ini' not found, ignore this if you're running first time
2016/04/25 15:26:22 [T] Custom path: /data/gogs
2016/04/25 15:26:22 [T] Log path: /app/gogs/log
2016/04/25 15:26:22 [I] Gogs: Go Git Service 0.8.42.0222
2016/04/25 15:26:22 [I] Log Mode: Console(Trace)
2016/04/25 15:26:22 [I] Cache Service Enabled
2016/04/25 15:26:22 [I] Session Service Enabled
2016/04/25 15:26:22 [I] SQLite3 Supported
2016/04/25 15:26:22 [I] Run Mode: Development
Apr 25 15:26:22 sshd[31]: Server listening on :: port 22.
Apr 25 15:26:22 sshd[31]: Server listening on 0.0.0.0 port 22.
2016/04/25 15:26:22 [I] Listen: http://0.0.0.0:3000
[Macaron] Started GET / for 172.17.0.1
2016/04/25 15:27:43 [D] Session ID: 06f9a9215839f3dc
2016/04/25 15:27:43 [D] CSRF Token: 56XTN2OhbC1q3j4SXq_DWAC2BJA6MTQ2MTU5ODA2MzU4NjE0MDgyMQ==
[Macaron] Completed / 302 Found in 685.67µs
[Macaron] Started GET /install for 172.17.0.1
2016/04/25 15:27:43 [D] Session ID: 06f9a9215839f3dc
2016/04/25 15:27:43 [D] CSRF Token: 56XTN2OhbC1q3j4SXq_DWAC2BJA6MTQ2MTU5ODA2MzU4NjE0MDgyMQ==
2016/04/25 15:27:43 [D] Template: install
[Macaron] Completed /install 200 OK in 62.764508ms
[Macaron] Started GET /js/jquery-1.11.3.min.js for 172.17.0.1
[Macaron] [Static] Serving /js/jquery-1.11.3.min.js
[Macaron] Completed /js/jquery-1.11.3.min.js 200 OK in 301.552µs
[Macaron] Started GET /assets/font-awesome-4.5.0/css/font-awesome.min.css for 172.17.0.1
[Macaron] [Static] Serving /assets/font-awesome-4.5.0/css/font-awesome.min.css
[Macaron] Completed /assets/font-awesome-4.5.0/css/font-awesome.min.css 200 OK in 193.589µs
[Macaron] Started GET /assets/octicons-3.4.1/octicons.css for 172.17.0.1
[Macaron] [Static] Serving /assets/octicons-3.4.1/octicons.css
[Macaron] Completed /assets/octicons-3.4.1/octicons.css 200 OK in 177.526µs
[Macaron] Started GET /css/semantic-2.1.8.min.css for 172.17.0.1
[Macaron] [Static] Serving /css/semantic-2.1.8.min.css
[Macaron] Completed /css/semantic-2.1.8.min.css 200 OK in 1.254994ms
[Macaron] Started GET /css/gogs.css?v=49fa7bd44f41088b9d40bac3f8a7941d for 172.17.0.1
[Macaron] [Static] Serving /css/gogs.css
[Macaron] Completed /css/gogs.css?v=49fa7bd44f41088b9d40bac3f8a7941d 200 OK in 280.138µs
[Macaron] Started GET /js/semantic-2.1.8.min.js for 172.17.0.1
[Macaron] [Static] Serving /js/semantic-2.1.8.min.js
[Macaron] Completed /js/semantic-2.1.8.min.js 200 OK in 491.494µs
[Macaron] Started GET /js/gogs.js?v=49fa7bd44f41088b9d40bac3f8a7941d for 172.17.0.1
[Macaron] [Static] Serving /js/gogs.js
[Macaron] Completed /js/gogs.js?v=49fa7bd44f41088b9d40bac3f8a7941d 200 OK in 239.216µs
[Macaron] Started GET /js/libs/emojify-1.1.0.min.js for 172.17.0.1
[Macaron] [Static] Serving /js/libs/emojify-1.1.0.min.js
[Macaron] Completed /js/libs/emojify-1.1.0.min.js 200 OK in 189.947µs
[Macaron] Started GET /js/libs/clipboard-1.5.5.min.js for 172.17.0.1
[Macaron] [Static] Serving /js/libs/clipboard-1.5.5.min.js
[Macaron] Completed /js/libs/clipboard-1.5.5.min.js 200 OK in 147.923µs
[Macaron] Started GET /css/themes/default/assets/fonts/icons.woff2 for 172.17.0.1
[Macaron] [Static] Serving /css/themes/default/assets/fonts/icons.woff2
[Macaron] Completed /css/themes/default/assets/fonts/icons.woff2 200 OK in 245.773µs
[Macaron] Started GET /img/favicon.png for 172.17.0.1
[Macaron] [Static] Serving /img/favicon.png
[Macaron] Completed /img/favicon.png 200 OK in 175.273µs
[Macaron] Started POST /install for 172.17.0.1
2016/04/25 15:36:49 [D] Session ID: 06f9a9215839f3dc
2016/04/25 15:36:49 [D] CSRF Token: 56XTN2OhbC1q3j4SXq_DWAC2BJA6MTQ2MTU5ODA2MzU4NjE0MDgyMQ==
2016/04/25 15:37:00 [T] Custom path: /data/gogs
2016/04/25 15:37:00 [T] Log path: /app/gogs/log
2016/04/25 15:37:00 [I] Gogs: Go Git Service 0.8.42.0222
2016/04/25 15:37:00 [I] Log Mode: File(Info)
2016/04/25 15:37:00 [I] Cache Service Enabled
2016/04/25 15:37:00 [I] Session Service Enabled
2016/04/25 15:37:00 [I] Git Version: 2.6.4
2016/04/25 15:37:00 [I] Git config user.name set to Gogs
2016/04/25 15:37:00 [I] Git config user.email set to gogs@fake.local
2016/04/25 15:37:00 [T] Doing: CheckRepoStats
2016/04/25 15:37:00 [I] SQLite3 Supported
2016/04/25 15:37:00 [I] Run Mode: Production
2016/04/25 15:37:00 [I] First-time run install finished!
[Macaron] Completed /install 302 Found in 10.67268627s
```

So are the setup directions wrong or is there a bug in gogs?

**Comment by oleersoy** [comment]
I think the application url instruction is wrong.  It says this:

> Application URL: Use combination of Domain and exposed HTTP Port values (e.g. http://192.168.99.100:10080/).

This does not work because it's combining the exposed port with the docker container IP.

When accessing gogs from the host that the docker container is running on we either have to go through localhost like this:
`localhost:10080`

Or directly through the docker virtual host like this:
`172.17.0.3:3000`

So I think the last setup instruction should work like be stated like this:
> Application URL: Use combination of Domain (Docker container IP) and the container HTTP Port (3000 - e.g. http://192.168.99.100:3000/).

**Comment by Unknwon** [comment]
>I think the application url instruction is wrong. It says this:

3000 is the port inside Docker, and you mapping 10080 to 3000 which mean for outside world, they only know 10080.

For the IP address, I think it's your config/understanding problem of Docker.

**Comment by oleersoy** [comment]
The instructions the way they are now will always lead the user to a failed install.  The reason is that Gogs will attempt to complete the install on the port number that is the external host port number, instead of the internal number.  So for example this will work:

http://192.168.99.100:3000/

Or this:
http://127.0.0.1:10080/
http://localhost:10080/

So the recommendation needs to be one of the two.  The two are different usage scenarios with the latter being on where Gogs is being accessed by the outside world through port forwarding, and the first one being one where Gogs is accessed locally (On a laptop) like I'm doing.  The first option is the simplest one to follow.  If the user is starting with that option they can skip port mapping and just start the container like this (Uses a data volume):

`docker run --name=gogs -v gogs-data:/data gogs/gogs`

**Comment by oleersoy** [comment]
BTW - Thanks for creating Gogs - to say I'm impressed would be an extreme understatement.

**Comment by Unknwon** [comment]
>The instructions the way they are now will always lead the user to a failed install.

No, the Docker README says very clear where to port 3000, and where to put 10080.

**Comment by oleersoy** [comment]
True - but the instructions are wrong.  They mix port 10080 with the Gogs container IP.  Port 10080 should go with the host running the docker container, and port 3000 can be used when accessing the docker container as if it were an external host.

**Comment by Unknwon** [comment]
The host IP is not always the Docker container IP, and it normally should not be.

cc/ @0rax

**Comment by oleersoy** [comment]
Absolutely true.  And if someone is an advanced user that is setting this up for team use and understands that, then they will be in good shape.  I was following the instructions as written up here:

https://github.com/gogits/gogs/tree/master/docker

And for domain it says:
> Domain: fill in with Docker container IP (e.g. 192.168.99.100)

And for the application URL it says this:
> Use combination of Domain and exposed HTTP Port values

So as you can see this mixes the docker container IP with the port exposed on the host (Which has a different public IP in this case).

**Comment by Unknwon** [comment]
I don't see any problem with these two descriptions, domain does not have a port, domain and a port becomes an URL. Docker container does not have domain name so uses its IP.

It's more like understanding problem, we should move discussion to https://discuss.gogs.io/

**Comment by oleersoy** [comment]
Sure - do you have a specific thread you would like to use.  It's really not complicated though.  In the instructions it says to set **domain** to the docker IP address.  So **domain** is the docker container virtual machine.  Trying to access that machine using the `exposed` port value, usually 10080, fails because the **domain** or docker container is listening on port 3000.

**Comment by Unknwon** [comment]
By `domain`, it is the domain/IP outside world could access to it, if you use internal IP inside container, that fails obviously...

**Comment by oleersoy** [comment]
That's what the instructions say to do:
> Domain: fill in with Docker container IP (e.g. 192.168.99.100)

So it's natural to think that the `domain` is the docker IP address ... 

I understand exactly what you mean and you understand exactly what you mean and it's obvious to us because we simply get it, but for other people trying to get this running for the first time it's easier to just follow step by step instructions.  I think that more people will fall into this trap unless the setup instructions are elaborated on.  Something like:

## Application URL
### Setup for World Access 
Set the domain to your hosts public / network accessible IP address and use the `10080` port that is mapped to the docker container port.  For example ....
### Access by Localhost Only
Set the domain to the docker IP address and use the internal container port `3000`.  For example ...

I actually think users will want neither of these, because now whenever they access gogs they have to use port numbers (The Git user can't access port 80).  So I think it would be good to mention that and explain the option:

### Use NGINX for clean URLs
I setup a separate NGINX container that proxies the docker container like this:
        server {
            listen 80;
            server_name 203.0.113.2;

            proxy_set_header X-Real-IP  $remote_addr; # pass on real client IP

            location / {
                proxy_pass http://203.0.113.1:3000;
            }
        }

I know that Gogs runs SSH internally ATM, but the whole thing might be cleaner with three containers:
1) NGINX Container for Proxying
2) Gogs (Pure - no SSH) Container
3) SSH Container

Manage all three the standard Docker way (I think swarm - still reading up).

**Comment by Unknwon** [comment]
`domain` could be anything that is possible to let outside world visit your Docker container.

**Comment by oleersoy** [comment]
I do understand.  I'm sure anyone Google'ing this now will end up here and can read the notes, so if it's OK with you I'll close this.

**Comment by Unknwon** [comment]
Yeah. You can close this one..

**Comment by arces** [comment]
Hello, I am having this same issue. It is only serving to the internal docker ip (accessible if I am on the local pc) but not accessible to outside pcs on the same network (or even the host pc using the host ip address). 
The gog's instance is on a network with a bridged connection so it can access the outside world.
I set up the domain and application url correctly too.
It worked just fine until I restarted it and then it didn't want to serve over that host ip address anymore.
